### PR TITLE
Modify the thread pool to adopt mkldnn models

### PR DESCRIPTION
**Pull request by i8run**
## What changes were proposed in this pull request?

This patch will modify the threadpools.

1. The `default` will support single thread including `LocalOptimizer` and `DistriOptimizer`. For supporting single thread version of `invokeWait2` method in `ThreadPool`, it will make the `threadpool` to current thread.
2. For dnn model, it will use the affinity to bind omp thread. And for performance issue, the `default` thread must use current main thread.
3. For `MTLabeldBGRImgToBatch` will use another new threads pool which is called `io`. So it will not be blocked when the `default` thread pool is single thread.
4. For `FileWriter`, it will not use `default`, otherwise the whole app will stuck at creating summary.

## How was this patch tested?

Manual tests and Jenkins.

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala [hunk 1]
```diff
@@ -41,6 +43,7 @@ class ThreadPool(private var poolSize: Int) {
 
   private def spawnThreadPool(poolSize: Int): ExecutionContext = {
     if (poolSize == 1) {
+      threadPool = MoreExecutors.sameThreadExecutor()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala]: looks like the threadPool is not used

Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala]: Looks like the sameThreadExecutor is deprecated?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 2]
```diff
@@ -515,4 +544,15 @@ object Engine {
       throw new IllegalArgumentException(s"Engine.init: Unsupported master format $master")
     }
   }
+
+  def setMklDnnEnvironments(): Unit = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: is this a private method?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 3]
```diff
@@ -515,4 +544,15 @@ object Engine {
...
+    if (System.getProperty("bigdl.engineType") == "mkldnn") { // do not use engineType
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: what do you mean about "do not use engineType" in the comment?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 4]
```diff
@@ -515,4 +544,15 @@ object Engine {
...
+      val threadsNumber = Math.ceil(Runtime.getRuntime.availableProcessors() / 2).toInt
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: what if the machine core number is 1

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 5]
```diff
@@ -327,6 +350,12 @@ object Engine {
       _model = new ThreadPool(modelPoolSize)
       _model.setMKLThread(MKL.getMklNumThreads)
     }
+
+    if (engineType == MklDnn) {
+      MKL.setNumThreads(MKL.getMklNumThreads)
+      com.intel.analytics.bigdl.mkl.MklDnn.setNumThreads(MKL.getMklNumThreads)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: why not import com.intel.analytics.bigdl.mkl.MklDnn?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 6]
```diff
@@ -310,11 +320,24 @@ object Engine {
...
+    }
+    _io
+  }
+
   private def initThreadPool(core : Int) : Unit = {
     val defaultPoolSize: Int = System.getProperty("bigdl.utils.Engine.defaultPoolSize",
       (core * 50).toString).toInt
     if(_default == null || _default.getPoolSize != defaultPoolSize) {
       _default = new ThreadPool(defaultPoolSize)
+      _default.setMKLThread(MKL.getMklNumThreads)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: will this affect the MKLBLAS model execution?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 7]
```diff
@@ -38,6 +39,12 @@ case object MklDnn extends EngineType
 
 
 object Engine {
+
+  // initialize some properties. Do not call it by manually.
+  // Because when we create some variables, they will set relevant properties, such as `_model`.
+  // If we do, any change of property for mkldnn will have no effect.
+  setMklDnnEnvironments()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: It looks a little confused. what about if(System.property("...") == mkldnn) setMklDnnEnvironments?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 8]
```diff
@@ -327,6 +350,12 @@ object Engine {
       _model = new ThreadPool(modelPoolSize)
       _model.setMKLThread(MKL.getMklNumThreads)
     }
+
+    if (engineType == MklDnn) {
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: The code is similar to what `setMKLThread` does for each thread; maybe put it into a method and then invoke the method

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala (hunk 1 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala]: `invokeAndWait2` will use `threadPool` and the `default` pool size maybe 1.

Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala]: The `sameThreadExecutor` is deprecated for new version of guava and is replaced with `directExecutor`. But we use 11.0.2 which has only the `sameThreadExecutor`.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala (hunk 6 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: It has no effects about the MKLBLAS model.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala (hunk 8 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: Yes. I'll move them to a new methods in `ThreadPool` object.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala (hunk 2 above)
Comment (i8run) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: Yes.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 9]
```diff
@@ -515,4 +538,15 @@ object Engine {
       throw new IllegalArgumentException(s"Engine.init: Unsupported master format $master")
     }
   }
+
+  private def setMklDnnEnvironments(): Unit = {
+    if (System.getProperty("bigdl.engineType") == "mkldnn") { // do not use Engine.engineType
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: why not put the if statement outside the function?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 10]
```diff
@@ -310,13 +319,25 @@ object Engine {
     _default
   }
 
+  private[bigdl] def io: ThreadPool = {
+    if (_io == null) {
+      throw new IllegalStateException(s"Engine.init: Thread engine is not " +
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: why not init _io here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala [hunk 11]
```diff
@@ -515,4 +538,15 @@ object Engine {
       throw new IllegalArgumentException(s"Engine.init: Unsupported master format $master")
     }
   }
+
+  private def setMklDnnEnvironments(): Unit = {
+    if (System.getProperty("bigdl.engineType") == "mkldnn") { // do not use Engine.engineType
+      val threadsNumber = Math.ceil(CpuInfo.getPhysicalProcessorCount).toInt
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala]: not / 2?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala (hunk 1 above)
Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ThreadPool.scala]: I think the code structure is weird. And we should not use a deprecated method

**Pull request merged**